src/tasks/block.rs

```diff
@@ -1,329 +1,352 @@
+    pub config: BuilderConfig,
+    /// A provider that cannot sign transactions, used for interacting with the rollup.
+    pub ru_provider: RuProvider,
+    /// The slot calculator for determining when to wake up and build blocks.
+    pub slot_calculator: SlotCalculator,
 }
 
-impl InProgressBlock {
-    /// Create a new `InProgressBlock`
-    pub const fn new() -> Self {
-        Self { transactions: Vec::new(), raw_encoding: OnceLock::new(), hash: OnceLock::new() }
-    }
-
-    /// Get the number of transactions in the block.
-    pub fn len(&self) -> usize {
-        self.transactions.len()
-    }
-
-    /// Check if the block is empty.
-    pub fn is_empty(&self) -> bool {
-        self.transactions.is_empty()
-    }
-
-    /// Unseal the block
-    fn unseal(&mut self) {
-        self.raw_encoding.take();
-        self.hash.take();
-    }
-
-    /// Seal the block by encoding the transactions and calculating the contentshash.
-    fn seal(&self) {
-        self.raw_encoding.get_or_init(|| encode_txns::<Alloy2718Coder>(&self.transactions).into());
-        self.hash.get_or_init(|| keccak256(self.raw_encoding.get().unwrap().as_ref()));
+type AlloyDatabaseProvider = WrapDatabaseAsync<AlloyDB<Ethereum, RuProvider>>;
+
+impl Simulator {
+    /// Creates a new `Simulator` instance.
+    ///
+    /// # Arguments
+    ///
+    /// - `config`: The configuration for the builder.
+    /// - `ru_provider`: A provider for interacting with the rollup.
+    /// - `slot_calculator`: A slot calculator for managing block timing.
+    ///
+    /// # Returns
+    ///
+    /// A new `Simulator` instance.
+    pub fn new(
+        config: &BuilderConfig,
+        ru_provider: RuProvider,
+        slot_calculator: SlotCalculator,
+    ) -> Self {
+        Self { config: config.clone(), ru_provider, slot_calculator }
     }
 
-    /// Ingest a transaction into the in-progress block. Fails
-    pub fn ingest_tx(&mut self, tx: &TxEnvelope) {
-        trace!(hash = %tx.tx_hash(), "ingesting tx");
-        self.unseal();
-        self.transactions.push(tx.clone());
+    /// Handles building a single block.
+    ///
+    /// # Arguments
+    ///
+    /// - `constants`: The system constants for the rollup.
+    /// - `sim_items`: The simulation cache containing transactions and bundles.
+    /// - `finish_by`: The deadline by which the block must be built.
+    ///
+    /// # Returns
+    ///
+    /// A `Result` containing the built block or an error.
+    pub async fn handle_build(
+        &self,
+        constants: SignetSystemConstants,
+        sim_items: SimCache,
+        finish_by: Instant,
+    ) -> Result<BuiltBlock, mpsc::error::SendError<BuiltBlock>> {
+        tracing::debug!(finish_by = ?finish_by, "starting block simulation");
+
+        let db = self.create_db().await.unwrap();
+        let block_build: BlockBuild<_, NoOpInspector> = BlockBuild::new(
+            db,
+            constants,
+            PecorinoCfg {},
+            NoopBlock,
```

Comment:
can't use a noop-block here. need to load a header and modify in a way similar to what happens here:

https://github.com/init4tech/signet-sdk/blob/main/crates/rpc/src/ctx.rs#L500-L523